### PR TITLE
[expo-notifications] Add useInitialNotificationResponse hook

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

May finally fix the https://github.com/expo/expo/issues/9866. Basically I have to admit API around initial notification response could have been a little bit better from the beginning (instead of "they'll figure it out" approach it should be "here's everything you need").

# How

Implemented a React hook and backend for it that returns:
- `undefined` until it knows whether to return:
- `null` if the app wasn't opened as a result of tapping on a notification
- `NotificationResponse` if the app was opened as a result of tapping on a notification.

# Test Plan

I have manually tested that the hook provides proper value on both iOS and Android. Actually it was easier on Android where the event comes in between `useLayoutEffect` and `useEffect`. On iOS `useEffect` was being executed earlier and the global listener was being cleared before it was triggered. I used `requestAnimationFrame`/`process.nextTick` polyfill to give it time.

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]
Looks good 🚀

File: docs/pages/versions/unversioned/sdk/notifications.md [hunk 1]
```diff
@@ -385,6 +386,75 @@ Removes all push token subscriptions that may have been registered with `addPush
 
 ## Listening to notification events
 
+### `useInitialNotificationResponse(): undefined | NotificationResponse | null`
+
+A React hook returning an initial notification response, i.e. a notification response responsible for opening the application (eg. tapping on a notification, triggering a notification action).
+
+#### Returns
+
+The hook may return values of three types/values:
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/notifications.md:395]:
```suggestion
The hook may return one of these three types/values:
```

**Review by cruzach — COMMENTED** [review]
That issue seems to have been really plaguing quite a few people, thank you so much 🙏  (personally, after placing the responseReceivedListener in the global scope I haven't been able to reproduce the bug)

Is this hook _only_ for listening to responses if the app is opened from a completely killed state, but for notification responses that occur while the app is open or backgrounded, people should still rely on `addNotificationResponseReceivedListener`?

**Comment by sjchmiela** [comment]
> Is this hook only for listening to responses if the app is opened from a completely killed state, but for notification responses that occur while the app is open or backgrounded, people should still rely on addNotificationResponseReceivedListener?

That is correct. Maybe apart from providing a `useInitialNotificationResponse()` hook we should provide a `useIncomingNotificationResponse()` that would register/unregister the listener inside too? It'd be more along the lines of what I wrote initially (with the even emitter passing new responses from global scope to mounted hooks), but without clearing the subscription up. I think `useinitialNotificationResponse` comes in handy when a developer wants to only render the application once they're sure that an app was or wasn't opened after a notification tap (either because it's how they developed their navigation — they need this information upfront or maybe the app renders so slowly that they want to wait 30ms for information than some other duration for app rerender [are those reasons even probable at all?]).

Do you think this hook makes sense or should I change it to `useIncomingNotificationResponse` (or some other better name)?

**Comment by cruzach** [comment]
> they want to wait 30ms for information...are those reasons even probable at all?

If I understand the theoretical situation correctly, yes- I know of at least a couple partners that make a large number of API calls on app startup that can take up to 500ms (sometimes even longer)

> `useinitialNotificationResponse` comes in handy when a developer wants to only render the application once they're sure that an app was or wasn't opened after a notification tap

I agree that this is a really nice feature of this hook as it currently is. If a new hook (like `useIncomingNotificationResponse`) _could_ do that **and** pass any new responses while the app is running as well, then of course that would be the ideal situation- one hook/function to rely on for all my notification response needs. **But**, if we can't have an "all-in-one" hook, it's good to just document that this hook **only** applies to incoming responses when the app is killed, and for other cases you should rely on the `notificationResponseReceivedListener`.

btw `useIncomingNotificationResponse` is a perfect name imo, can't think of anything clearer than that

File: packages/expo-notifications/src/useInitialNotificationResponse.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export default function useInitialNotificationResponse() {
+  const [initialNotificationResponse, setInitialNotificationResponse] = useState<
+    NotificationResponse | null | undefined
+  >(globalInitialNotificationResponse);
+
+  useEffect(() => {
+    // process.nextTick & requestAnimationFrame-like,
+    // without this on iOS the subscription is cleared
+    // before it's triggered
+    setTimeout(() => {
```
Comment (SimenB) [inline packages/expo-notifications/src/useInitialNotificationResponse.ts:53]: no cleanup?

Comment (sjchmiela) [reply packages/expo-notifications/src/useInitialNotificationResponse.ts:53]: I've had some problems with the hook with the cleanup added and the `setTimeout(…, 0)` is almost immediate so it didn't feel like it needs cleanup, but I'll try it out again with new implementations, thanks for ensuring it doesn't pass without superproper attention 👍

**Review by cruzach — APPROVED** [review]

**Pull request merged**